gadget/filesystemimage.go

```diff
@@ -0,0 +1,133 @@
+	}
+)
+
+// FilesystemImageWriter is capable of creating filesystem images described by
+// positioned structures.
+type FilesystemImageWriter struct {
+	contentDir string
+	ps         *PositionedStructure
+	workDir    string
+}
+
+// PostStageFunc is called after the filesystem contents for the given structure
+// have been staged at a temporary location
+type PostStageFunc func(rootDir string, ps *PositionedStructure) error
+
+// Write creates the filesystem inside the provided image file and populates it
+// with data according to content declartion of the structure. Content data is
+// staged in a temporary location. An optional post-stage helper function can be
+// used to manipulate the data before it is copied over to the image.
+func (f *FilesystemImageWriter) Write(fname string, postStage PostStageFunc) error {
+	st, err := os.Stat(fname)
+	if err != nil {
+		return fmt.Errorf("cannot stat image file: %v", err)
+	}
+	if sz := st.Size(); sz != int64(f.ps.Size) {
+		return fmt.Errorf("size of image file %v is different from declared structure size %v", sz, f.ps.Size)
+	}
+
+	mkfs := mkfsHandlers[f.ps.Filesystem]
+	if mkfs == nil {
+		return fmt.Errorf("internal error: filesystem %q has no handler", f.ps.Filesystem)
+	}
+
+	where := filepath.Join(f.workDir, fmt.Sprintf("snap-stage-content-part-%04d", f.ps.Index))
+	if osutil.IsDirectory(where) {
+		return fmt.Errorf("cannot prepare staging directory %s: path exists", where)
+	}
+
+	if err := os.Mkdir(where, 0755); err != nil {
+		return fmt.Errorf("cannot prepare staging directory: %v", err)
+	}
+
+	if os.Getenv("SNAP_DEBUG_IMAGE_NO_CLEANUP") == "" {
+		defer func() {
+			if err := os.RemoveAll(where); err != nil {
+				logger.Noticef("cannot remove filesystem staging directory %q: %v", where, err)
+			}
+		}()
+	}
+
+	mrw, err := NewMountedFilesystemWriter(f.contentDir, f.ps)
+	if err != nil {
+		return fmt.Errorf("cannot prepare filesystem writer for %v: %v", f.ps, err)
+	}
+	// drop all contents to the staging directory
+	if err := mrw.Write(where, nil); err != nil {
+		return fmt.Errorf("cannot prepare filesystem content: %v", err)
+	}
+
+	if postStage != nil {
+		if err := postStage(where, f.ps); err != nil {
+			return fmt.Errorf("post stage callback failed: %v", err)
+		}
+	}
+
+	if err := mkfs(fname, f.ps.Label, where); err != nil {
+		return fmt.Errorf("cannot create %q filesystem: %v", f.ps.Filesystem, err)
+	}
+
+	return nil
+}
+
+// NewFilesystemImageWriter returns a writer capable of creating filesystem
+// images corresponding to the provided positiioned structure, with content from
+// the given content directory. A staging directory will be created in
+// optionally prided work directory, otherwise the default temporary storage
+// location will be used.
+func NewFilesystemImageWriter(contentDir string, ps *PositionedStructure, workDir string) (*FilesystemImageWriter, error) {
+	if ps == nil {
+		return nil, fmt.Errorf("internal error: *PositionedStructure is nil")
```

Comment:
No strong opinion here, but I think we haven't been using a "low level" message like this even for internal errors. Perhaps just "internal error: positioned structure cannot be nil" (based on existing: "internal error: assertion type cannot be nil").

Reply:
I'd keep the error, it's in sync with errors returned by other ctors: with https://github.com/snapcore/snapd/blob/master/gadget/raw.go#L42-L44